[JSC] Extract MicrotaskQueue

Constellation opened the pull request:
#### 43c25c9714e7fb514a0a2c70646dc457ab199ab7
<pre>
[JSC] Extract MicrotaskQueue
<a href="https://bugs.webkit.org/show_bug.cgi?id=289074">https://bugs.webkit.org/show_bug.cgi?id=289074</a>
<a href="https://rdar.apple.com/146088353">rdar://146088353</a>

Reviewed by Yijia Huang.

This patch extracts JSC&apos;s MicrotaskQueue to prepare to extend it to
fully cover WebCore use case. Right now it is only used for
JavaScriptCore framework etc.

1. This patch extends MicrotaskQueue to accept multiple instances. We
   have a linked list in VM and managing all the instances of
   MicrotaskQueue associated with this VM.
2. Move MicrotaskQueue to its own file.

* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/runtime/MicrotaskQueue.cpp: Added.
(JSC::QueuedTask::run):
(JSC::MicrotaskQueue::create):
(JSC::MicrotaskQueue::visitAggregateImpl):
* Source/JavaScriptCore/runtime/MicrotaskQueue.h: Added.
(JSC::QueuedTask::QueuedTask):
(JSC::QueuedTask::identifier const):
* Source/JavaScriptCore/runtime/VM.cpp:
(JSC::VM::VM):
(JSC::VM::queueMicrotask):
(JSC::VM::drainMicrotasks):
(JSC::VM::beginMarking):
(JSC::VM::visitAggregateImpl):
(JSC::QueuedTask::run): Deleted.
(JSC::MicrotaskQueue::visitAggregateImpl): Deleted.
* Source/JavaScriptCore/runtime/VM.h:
(JSC::QueuedTask::QueuedTask): Deleted.
(JSC::QueuedTask::identifier const): Deleted.
(JSC::MicrotaskQueue::dequeue): Deleted.
(JSC::MicrotaskQueue::enqueue): Deleted.
(JSC::MicrotaskQueue::isEmpty const): Deleted.
(JSC::MicrotaskQueue::size const): Deleted.
(JSC::MicrotaskQueue::clear): Deleted.
(JSC::MicrotaskQueue::beginMarking): Deleted.
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/291566@main">https://commits.webkit.org/291566@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5fce828eb2edf1660ffbd3a11c34951583a36e2f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/93314 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/12872 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/2600 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/98313 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/43840 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/95364 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/13159 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/21328 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/71313 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/28707 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/96316 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/9878 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/84406 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/51646 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/9567 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/2039 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/43154 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/86024 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/79853 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/2053 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/100345 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/91980 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/20366 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/14911 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/80332 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/20618 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/80326 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/79647 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/24194 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/1527 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/13491 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/14934 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/20350 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/25527 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/114630 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/20037 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/33059 "Found 14 new JSC stress test failures: microbenchmarks/memcpy-wasm-small.js.default, microbenchmarks/memcpy-wasm-small.js.dfg-eager, microbenchmarks/memcpy-wasm-small.js.mini-mode, stress/sampling-profiler-display-name.js.dfg-eager-no-cjit-validate, wasm.yaml/wasm/stress/b3-signed-extend-16-to-64.js.wasm-eager-jettison, wasm.yaml/wasm/stress/ipint-bbq-osr-with-try5.js.wasm-eager, wasm.yaml/wasm/stress/ipint-bbq-osr-with-try5.js.wasm-slow-memory, wasm.yaml/wasm/stress/live-funcref-across-loop-tier-up.js.default-wasm, wasm.yaml/wasm/stress/repro_1289.js.default-wasm, wasm.yaml/wasm/stress/simple-inline-exception-inlinee-catch-with-delegate.js.wasm-eager ... (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/23497 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/21778 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

MenloDorian commented on Source/JavaScriptCore/runtime/VM.h:
```diff
@@ -1154,6 +1086,7 @@ class VM : public ThreadSafeRefCountedWithSuppressingSaferCPPChecking<VM>, publi
     Lock m_loopHintExecutionCountLock;
     UncheckedKeyHashMap<const JSInstruction*, std::pair<unsigned, std::unique_ptr<uintptr_t>>> m_loopHintExecutionCounts;
 
+    Ref<MicrotaskQueue> m_defaultMicrotaskQueue;
```
What's the difference between `m_defaultMicrotaskQueue` and `m_microtaskQueues`?  What is the meaning of "default" in this case?

Constellation replied:
`default`-initialized queue and used when no specific MicrotaskQueue is instantiated. This is just keeping the current behavior for JSC framework etc. But later, we will add a feature having multiple MicrotaskQueue created at runtime for WebCore, then this is not used in the main thread (while WebCore Worklet still uses this default queue too).

hyjorc1 reviewed (approved):
r=me

Constellation committed:
[JSC] Extract MicrotaskQueue
https://bugs.webkit.org/show_bug.cgi?id=289074
rdar://146088353

Reviewed by Yijia Huang.

This patch extracts JSC's MicrotaskQueue to prepare to extend it to
fully cover WebCore use case. Right now it is only used for
JavaScriptCore framework etc.

1. This patch extends MicrotaskQueue to accept multiple instances. We
   have a linked list in VM and managing all the instances of
   MicrotaskQueue associated with this VM.
2. Move MicrotaskQueue to its own file.

* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/runtime/MicrotaskQueue.cpp: Added.
(JSC::QueuedTask::run):
(JSC::MicrotaskQueue::create):
(JSC::MicrotaskQueue::visitAggregateImpl):
* Source/JavaScriptCore/runtime/MicrotaskQueue.h: Added.
(JSC::QueuedTask::QueuedTask):
(JSC::QueuedTask::identifier const):
* Source/JavaScriptCore/runtime/VM.cpp:
(JSC::VM::VM):
(JSC::VM::queueMicrotask):
(JSC::VM::drainMicrotasks):
(JSC::VM::beginMarking):
(JSC::VM::visitAggregateImpl):
(JSC::QueuedTask::run): Deleted.
(JSC::MicrotaskQueue::visitAggregateImpl): Deleted.
* Source/JavaScriptCore/runtime/VM.h:
(JSC::QueuedTask::QueuedTask): Deleted.
(JSC::QueuedTask::identifier const): Deleted.
(JSC::MicrotaskQueue::dequeue): Deleted.
(JSC::MicrotaskQueue::enqueue): Deleted.
(JSC::MicrotaskQueue::isEmpty const): Deleted.
(JSC::MicrotaskQueue::size const): Deleted.
(JSC::MicrotaskQueue::clear): Deleted.
(JSC::MicrotaskQueue::beginMarking): Deleted.
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/291566@main
